src/libraries/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs

```diff
@@ -133,9 +141,10 @@ public static class LazyInitializer
         /// <param name="initialized">A reference to a boolean that determines whether the target has already
         /// been initialized.</param>
         /// <param name="syncLock">A reference to an object used as the mutually exclusive lock for initializing
-        /// <paramref name="target"/>. If <paramref name="syncLock"/> is null, a new object will be instantiated.</param>
+        /// <paramref name="target"/>. If <paramref name="syncLock"/> is null, and if the target hasn't already
+        /// been initialized, a new object will be instantiated.</param>
         /// <returns>The initialized value of type <typeparamref name="T"/>.</returns>
-        public static T EnsureInitialized<T>([AllowNull] ref T target, ref bool initialized, [NotNull] ref object? syncLock)
+        public static T EnsureInitialized<T>([AllowNull] ref T target, ref bool initialized, [NotNullIfNotNull("syncLock")] ref object? syncLock)
```

Comment:
Hmm.  This `[NotNullIfNotNull("syncLock")]` on the `syncLock` parameter is basically saying "the method won't set this ref parameter to null".  I wonder if we have other places where we should be adding such an attribute... to my knowledge these if the first time we've used `NotNullIfNotNull` to refer to the same parameter it's attributed on.

Reply:
For example, everywhere we have a `ref T foo` parameter, should we be annotating it as `[NotNullIfNotNull("foo")]`?  That'd be kind of awful.  @jcouv, what's the recommended approach here?

Reply:
Along those lines, `Volatile.Read<T>(ref T value)` could be a candidate for an `[In]`-like attribute. Basically: "The caller must explicitly use the _ref_ keyword to pass an address to this routine (no implicit _in_ conversion allowed), but the routine promises not to assign a new value to the provided address." That attribute would also imply `[NotNullIfNotNull("sameParam")]` by construction.